apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/SessionPoolTests.cs

```diff
@@ -0,0 +1,443 @@
+                await pool.CreateSessionFromPoolAsync(
+                    client.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+
+                var releaseTask = Task.Delay(TimeSpan.FromMilliseconds(10)).ContinueWith(
+                    (x, y) => { pool.ReleaseToPool(client.Object, session1); }, null);
+
+                var createTask = pool.CreateSessionFromPoolAsync(
+                    client.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None);
+
+                await Task.WhenAll(createTask, releaseTask).ConfigureAwait(false);
+                Assert.NotNull(createTask.ResultWithUnwrappedExceptions());
+            }
+        }
+
+        [Fact]
+        public async Task ClientsHaveDifferentPools()
+        {
+            using (var pool = new SessionPool())
+            {
+
+                var client1 = CreateMockClient();
+                var client2 = CreateMockClient();
+
+                var session = await pool.CreateSessionFromPoolAsync(
+                    client1.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                pool.ReleaseToPool(client1.Object, session);
+
+                var session2 = await pool.CreateSessionFromPoolAsync(
+                    client2.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+
+                Assert.NotSame(session, session2);
+                Assert.Equal(2, _createdSessions.Count);
+                Assert.Equal(1, _createdSessions[client1].Count);
+                Assert.Equal(1, _createdSessions[client2].Count);
+            }
+        }
+
+        [Fact]
+        public async Task DatabasesHaveDifferentPools()
+        {
+            using (var pool = new SessionPool())
+            {
+                var client = CreateMockClient();
+
+                var session = await pool.CreateSessionFromPoolAsync(
+                    client.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                pool.ReleaseToPool(client.Object, session);
+
+                var session2 = await pool.CreateSessionFromPoolAsync(
+                    client.Object, "newproject", "newinstance", "newdbid", null, CancellationToken.None)
+                    .ConfigureAwait(false);
+
+                Assert.NotSame(session, session2);
+                Assert.Equal(1, _createdSessions.Count);
+                Assert.Equal(1, _createdSessions[client].Count);
```

Comment:
because I am only tracking default database sessions up in CreateMockClient:
Key line is:
"                        It.Is<DatabaseName>(y => y.Equals(dbName)),  "
Could probably at some point add more tracking for other sessions.